Examples/GenerativeAICLI/Sources/GenerateContent.swift

```diff
@@ -93,27 +124,71 @@ struct GenerateContent: AsyncParsableCommand {
         parts.append(.data(mimetype: mimeType, imageData))
       }
 
-      let input = [ModelContent(parts: parts)]
+      input = [ModelContent(parts: parts)]
+
+      repeat {
+        try await processFunctionCalls()
 
-      if isStreaming {
-        let contentStream = model.generateContentStream(input)
-        print("Generated Content <streaming>:")
-        for try await content in contentStream {
-          if let text = content.text {
-            print(text)
+        if isStreaming {
+          let contentStream = model.generateContentStream(input)
+          print("Generated Content <streaming>:")
+          for try await content in contentStream {
+            processResponseContent(content: content)
           }
+        } else {
+          // Unary generate content
+          let content = try await model.generateContent(input)
+          print("Generated Content:")
+          processResponseContent(content: content)
         }
-      } else {
-        let content = try await model.generateContent(input)
-        if let text = content.text {
-          print("Generated Content:\n\(text)")
-        }
-      }
+      } while !functionCalls.isEmpty
     } catch {
       print("Generate Content Error: \(error)")
     }
   }
 
+  mutating func processResponseContent(content: GenerateContentResponse) {
```

Comment:
This is a lot of code that developers have to write, and I imagine it would look pretty much the same for most apps. Which is to say that I would expect handling of function calls to be part of the framework.

I'd expect either `generateContent` and `sendMessage` (and their respective streaming versions) or the model itself to provide a hook for a callback that developers can register to perform any validation / confirmation.

E.g.

enum FunctionCallInvocationDecision {
  case granted
  case rejected(reason: String)
}

func onBeforeFunctionCall(invocation: FunctionCall) -> FunctionCallInvocationDecision {
  // inspect function call

  // for any potentially harmful actions, ask user for confirmation if they want to execute the function
  if invocation.name == "book_flight"  {
    // ask user
   return .rejected("Flight too expensive")  
  }
}

Reply:
Invoking the functions on behalf of the developer is generally referred to as automatic function calling. It's also rough but #118 shows what I was working on towards that goal.

It's unclear to me whether we are required to support both manual and automatic function calling. In the example you provided (and in #118), it's possible for devs to not invoke the underlying function (returning `.rejected` in this example or throwing/returning some error response in mine).

I have seen some codelabs/examples where the function's response isn't returned back to the model. That might be one case where the manual approach could be preferred (but unclear).

Reply:
Will look into this further in follow-ups.